src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1482 @@
+
+    static ref DD_ENABLED: bool = env::var_os("IPC_CHANNEL_WIN_DEBUG_DUMP").is_some();
+    static ref DD2_ENABLED: bool = env::var_os("IPC_CHANNEL_WIN_DEBUG_MORE_DUMP").is_some();
+}
+
+macro_rules! dd { ($($rest:tt)*) => { if *DD_ENABLED { println!($($rest)*); } } }
+macro_rules! dd2 { ($($rest:tt)*) => { if *DD_ENABLED || *DD2_ENABLED { println!($($rest)*); } } }
+
+// We stick these in WinErrors to indicate various things
+// that aren't quite errors.
+const MAGIC_CODE_BASE: u32 = 0xdeadbee0u32;
+const MAGIC_NO_OUTSTANDING_DATA_CODE: u32 = MAGIC_CODE_BASE + 0;
+const MAGIC_CHANNEL_CLOSED_CODE: u32 = MAGIC_CODE_BASE + 1;
+
+// When we create the pipe, how big of a write buffer do we specify?
+// This is reserved in the nonpaged pool.  The fragment size is the
+// max we can write to the pipe without fragmentation, and the
+// buffer size is what we tell the pipe it is, so we have room
+// for out of band data etc.  We can probably do way less than
+// 16k extra.
+const MAX_FRAGMENT_SIZE: usize = 64 * 1024;
+const WRITE_BUFFER_SIZE: usize = MAX_FRAGMENT_SIZE + 16*1024;
+
+// When we read, how big of a heap buffer do we allocate?
+const READ_BUFFER_SIZE: usize = 128 * 1024; //128k;
+const READ_BUFFER_MAX_GROWTH: usize = 1 * 1024 * 1024; // 1MB max growth
+
+// The message header is 2xu32 -- first one is the length of the data,
+// second one is the length of any out of band data.  If the first
+// field is INVALID_HEADER_DATA_SIZE, then the second one signifies
+// a control code (currently only MAGIC_CHANNEL_CLOSED_CODE is used)
+const HEADER_SIZE: usize = 8; //mem::size_of::<u32>()*2;
+const INVALID_HEADER_DATA_SIZE: u32 = 0xffffffffu32;
+
+#[allow(non_snake_case)]
+fn GetLastError() -> u32 {
+    unsafe {
+        kernel32::GetLastError()
+    }
+}
+
+pub fn channel() -> Result<(OsIpcSender, OsIpcReceiver),WinError> {
+    let mut receiver = try!(OsIpcReceiver::new());
+    let sender = try!(receiver.sender());
+    Ok((sender, receiver))
+}
+
+// If we have any channel handles or shmem segments,
+// then we'll send an OsIpcOutOfBandMessage after
+// the data message.
+#[derive(Debug)]
+struct OsIpcOutOfBandMessage {
+    channel_is_sender: Vec<bool>,
+    channel_handles: Vec<intptr_t>,
+    shmem_sizes: Vec<u64>,
+    shmem_handles: Vec<intptr_t>,
+    big_data_receiver_handle: intptr_t,
+}
+
+impl OsIpcOutOfBandMessage {
+    fn new() -> OsIpcOutOfBandMessage {
+        OsIpcOutOfBandMessage {
+            channel_is_sender: vec![],
+            channel_handles: vec![],
+            shmem_sizes: vec![],
+            shmem_handles: vec![],
+            big_data_receiver_handle: 0,
+        }
+    }
+
+    fn needs_to_be_sent(&self) -> bool {
+        !self.channel_handles.is_empty() ||
+        !self.shmem_handles.is_empty() ||
+        self.big_data_receiver_handle != 0
+    }
+
+    fn add_sender(&mut self, handle: HANDLE) {
+        self.channel_is_sender.push(true);
+        self.channel_handles.push(handle as intptr_t);
+    }
```

Comment:
You have setter methods like `add_sender()` etc. to populate the data structure, thus encapsulating the internal representation on that side; but on the receiver side, you are accessing the internal fields directly... You could avoid that by using some kind of iterator method I think.

Not a big deal really -- this asymmetry just feels weird to me. To be honest, personally I probably wouldn't bother with encapsulation here at all...